File: cracking_the_coding_interview_6th/chapter16/cpp/16_14_v01.cpp
#include <algorithm>
#include <fstream>
#include <string>
#include <vector>
using namespace std;
using ll = long long;

struct Coord {
  double x, y;

  bool operator<(const Coord &other) {
    if (x == other.x) return y < other.y;
    return x < other.x;
  }
};

struct Line {
  // ax + by + c = 0
  double a, b, c;
};

Line get_line(double dx1, double dy1, double x, double y) {
  if (dx1 == 0) return Line{1, 0, -x};
  // y = px + q
  double p = dy1 / dx1;
  double q = y - p * x;
  return Line{-p, 1, -q}; // -px + 1y + -q = 0
}

Line most_passing_line(vector<Coord> v) {
  if (v.size() == 1) return Line{1, 0, -v[0].x};
  sort(v.begin(), v.end());
  ll max_cnt = 0;
  Line res;
  for (ll i = 0; i < v.size() - 1; ++i) {
    for (ll j = i + 1; j < v.size(); ++j) {
      double dx1 = v[i].x - v[j].x;
      double dy1 = v[i].y - v[j].y;
      ll cnt = 2;
      for (ll k = j + 1; k < v.size(); ++k) {
        double dx2 = v[i].x - v[k].x;
        double dy2 = v[i].y - v[k].y;
        if (dx1 * dy2 == dx2 * dy1) ++cnt;
      }
      if (max_cnt < cnt) {
        max_cnt = cnt;
        res = get_line(dx1, dy1, v[i].x, v[i].y);
      }
    }
  }
  return res;
}

void solve(const string &fp) {
  printf("# %s\n", fp.c_str());
  ifstream ifs(fp);
  ll n;
  ifs >> n;
  vector<Coord> v(n);
  for (ll i = 0; i < n; ++i) {
    ifs >> v[i].x >> v[i].y;
  }
  Line res = most_passing_line(v);
  printf("%fx + %fy + %f = 0\n", res.a, res.b, res.c);
  puts("");
}

int main() {
  solve("../testcases/16_14/01.txt");
  solve("../testcases/16_14/02.txt");
  solve("../testcases/16_14/03.txt");
  solve("../testcases/16_14/04.txt");
  solve("../testcases/16_14/05.txt");
}

// # ../testcases/16_14/01.txt
// 1.000000x + 1.000000y + 4.000000 = 0

// # ../testcases/16_14/02.txt
// -0.200000x + 1.000000y + -0.400000 = 0

// # ../testcases/16_14/03.txt
// 0.000000x + 1.000000y + -0.000000 = 0

// # ../testcases/16_14/04.txt
// 1.000000x + 0.000000y + -0.000000 = 0

// # ../testcases/16_14/05.txt
// 1.000000x + 0.000000y + -0.000000 = 0

File: cracking_the_coding_interview_6th/chapter16/python/16_14_v01.py
from __future__ import annotations
from typing import List


class Coord:
    def __init__(self, x: float, y: float):
        self.x: float = x
        self.y: float = y

    def __lt__(self, other: Coord):
        if self.x == other.x:
            if self.y < other.y:
                return -1
            elif self.y == other.y:
                return 0
            else:
                return 1
        if self.x < other.x:
            return -1
        else:
            return 1


class Line:
    # ax + by + c = 0
    def __init__(self, a: float, b: float, c: float):
        self.a: float = a
        self.b: float = b
        self.c: float = c


def get_line(dx1: float, dy1: float, x: float, y: float) -> Line:
    if dx1 == 0:
        return Line(1, 0, -x)
    # y = px + q
    p = dy1 / dx1
    q = y - p * x
    return Line(-p, 1, -q)


def most_passing_line(v: List[Coord]) -> Line:
    if len(v) == 1:
        return Line(1, 0, -v[0].x)
    v = sorted(v)
    max_cnt = 0
    n = len(v)
    for i in range(n - 1):
        for j in range(i + 1, n):
            dx1 = v[i].x - v[j].x
            dy1 = v[i].y - v[j].y
            cnt = 2
            for k in range(j + 1, n):
                dx2 = v[i].x - v[k].x
                dy2 = v[i].y - v[k].y
                if dx1 * dy2 == dy1 * dx2:
                    cnt += 1
            if max_cnt < cnt:
                max_cnt = cnt
                res = get_line(dx1, dy1, v[i].x, v[i].y)
    return res


def ns(f):
    return next(f).strip()


def solve(fp: str):
    print(f"# {fp}")
    with open(fp) as f:
        n = int(ns(f))
        v = [Coord(*map(float, ns(f).split())) for _ in range(n)]
    res = most_passing_line(v)
    print(f"{res.a}x + {res.b}y + {res.c} = 0")
    print()


solve("../testcases/16_14/01.txt")
solve("../testcases/16_14/02.txt")
solve("../testcases/16_14/03.txt")
solve("../testcases/16_14/04.txt")
solve("../testcases/16_14/05.txt")

# # ../testcases/16_14/01.txt
# 1.0x + 1y + 4.0 = 0

# # ../testcases/16_14/02.txt
# -0.2x + 1y + -0.3999999999999999 = 0

# # ../testcases/16_14/03.txt
# -0.0x + 1y + -0.0 = 0

# # ../testcases/16_14/04.txt
# 1x + 0y + -0.0 = 0

# # ../testcases/16_14/05.txt
# 1x + 0y + -0.0 = 0
